Fixed AngleDiff and AngleDist function definitions and added unit tests.

kylevedder opened the pull request.

kylevedder committed:
Updated AngleDiff and AngleDist to support differentiation for generic versions and added performant specializations for `double` and `float`. Added unit tests to verify correctness.

joydeep-b commented on math/geometry.h:
```diff
@@ -78,6 +78,23 @@ T GetNormOrZero(const Eigen::Matrix<T, N, 1>& vec) {
+T AngleDist(const T a1, const T a2) {
+  const auto h1 = Heading(a1);
+  const auto h2 = Heading(a2);
+  const auto angle = std::acos(h1.dot(h2));
```
Please use `atan2`, not acos. The co-domain of acos is not the full -pi to pi.

joydeep-b commented on math/geometry.h:
```diff
@@ -78,6 +78,23 @@ T GetNormOrZero(const Eigen::Matrix<T, N, 1>& vec) {
+T AngleDiff(const T a1, const T a2) {
+  const auto h1 = Heading(a1);
+  const auto h2 = Heading(a2);
+  const auto angle = std::acos(h1.dot(h2));
```
Please use `atan2`, not acos. The co-domain of acos is not the full -pi to pi.

joydeep-b commented on math/geometry.h:
```diff
@@ -78,6 +78,23 @@ T GetNormOrZero(const Eigen::Matrix<T, N, 1>& vec) {
+  const auto h1 = Heading(a1);
+  const auto h2 = Heading(a2);
+  const auto angle = std::acos(h1.dot(h2));
+  const int sign = math_util::Sign(Cross(h1, h2));
```
You won't need to do this with `atan2`.

joydeep-b reviewed (changes requested):
Thanks for the contributions, @kylevedder .
I separately made this change on my end to be auto-diff friendly, but it would be good to have it in the main library.
Two main changes needed:
1. Please use `atan2`, not `acos`.
2. Please add in the template specializations for `float` and `double` using `rint` instead of the trig functions: it will be **much** faster for them.

kylevedder commented:
1. Every `atan2` based approach I can think of has numerical stability problems (e.g. when the two angles are the same) and I don't see why it'd be any faster. What was the exact mathematical formula you were looking for? I also don't understand the concern over `acos`'s co-domain. `AngleDist(a, b)` measures the unsigned distance between `a` and `b` which will fall from `0` to `pi`, so the naked use of `acos` is fine. `AngleDiff` measure the signed distance between `a` and `b` which will fall from `-pi` to `pi`, but the sign computation takes care of negating the `acos`'s output properly so it will on the full range from `-pi` to `pi`.

2. I don't understand what you are looking for with the template specializations for `float` and `double` using `rint` specializations; `rint` rounds floating point numbers to integral values but I don't understand why that's relevant to these functions.

I added another test to check that the case of `-pi` to `pi` is correctly returned as `0`.

joydeep-b commented:
@kylevedder 
There seem to be a number of differences in how you expect these functions to work, and how we do. Let's talk on Zoom to go over them, it'll be faster.

joydeep-b commented:
Correct AngleDiff implementation:

```cpp
template <typename T>
T AngleDiff(T a0, T a1) {
  const T s = std::sin(a0 - a1);
  const T c = std::cos(a0 - a1);
  return std::atan2(s, c);
}
```

kylevedder commented:
I have moved the functions back to `math_util`, added a generic version that uses trig functions along with template specialization for `float` and `double`, and added tests which check the generic trig version, the template specialization of `float`, and the template specialization of `double`.

joydeep-b commented on math/math_util.h:
```diff
@@ -59,14 +59,34 @@ T AngleMod(T angle) {
 
 template <typename T>
 T AngleDiff(T a0, T a1) {
```
This should be `const T&`.

kylevedder replied:
Fixed this and all other relevant functions.

joydeep-b reviewed (changes requested):
One small change needed.

joydeep-b reviewed (approved):
LGTM. Please merge when ready.
